src/Rewards.sol

```diff
@@ -159,30 +184,78 @@ contract Rewards is AccessControl, EIP712 {
      * @param signature The signature from the authorized oracle.
      */
     function mintReward(Reward memory reward, bytes memory signature) external {
+        _mustBeExpectedSequence(reward.recipient, reward.sequence);
+
         _mustBeFromAuthorizedOracle(digestReward(reward), signature);
 
-        _mustBeExpectedSequence(reward.recipient, reward.sequence);
+        _checkedUpdateQuota();
+
+        _checkedUpdateClaimed(reward.amount);
+
+        // Safe to increment the sequence after checking this is the expected number (no overflow for the age of universe even with 1000 reward claims per second)
+        sequences[reward.recipient] = reward.sequence + 1;
+
+        nodl.mint(reward.recipient, reward.amount);
+
+        emit Minted(reward.recipient, reward.amount, claimed);
+    }
+
+    /**
+     * @dev Mints batch rewards to multiple recipients.
+     * @param batch The BatchReward struct containing the recipients and amounts of rewards to be minted.
+     * @param signature The signature to verify the authenticity of the batch reward.
+     */
+    function mintBatchReward(BatchReward memory batch, bytes memory signature) external {
+        _mustBeValidBatchStructure(batch);
+
+        _mustBeExpectedBatchSequence(batch.sequence);
+
+        _mustBeFromAuthorizedOracle(digestBatchReward(batch), signature);
+
+        _checkedUpdateQuota();
 
+        uint256 batchSum = _batchSum(batch);
+
+        _checkedUpdateClaimed(batchSum);
+
+        // Safe to increment the sequence after checking this is the expected number (no overflow for the age of universe even with 1000 reward claims per second)
+        batchSequence = batch.sequence + 1;
+
+        for (uint256 i = 0; i < batch.recipients.length; i++) {
+            nodl.mint(batch.recipients[i], batch.amounts[i]);
+        }
```

Comment:
We can avoid the double for loop here (other loop in `_batchSum`) by rewriting the for iteration and add one last revert statement afterwards

Reply:
I tested the potential gas saving in your suggestion and found it either absolute zero or very negligible (less than 0.1%). While checking first and minting later has the benefit of early failure (gas saving there) and follows the best practice of updating state first before calling into any other contract.

Reply:
Makes sense, let's keep it as is